techlibs/quicklogic/synth_quicklogic.cc

```diff
@@ -0,0 +1,150 @@
+        log("\n");
+        log("    -edif <file>\n");
+        log("        write the design to the specified edif file. writing of an output file\n");
+        log("        is omitted if this parameter is not specified.\n");
+        log("\n");
+        log("    -blif <file>\n");
+        log("        write the design to the specified BLIF file. writing of an output file\n");
+        log("        is omitted if this parameter is not specified.\n");
+        log("\n");
+        log("    -flatten\n");
+        log("        flatten design before synthesis\n");
+        log("\n");
+        help_script();
+        log("\n");
+    }
+
+    std::string top_opt = "";
+    std::string edif_file = "";
+    std::string blif_file = "";
+    std::string currmodule = "";
+    bool flatten = false;
+
+    void clear_flags() YS_OVERRIDE
+    {
+        top_opt = "-auto-top";
+        flatten = false;
+        edif_file.clear();
+        blif_file.clear();
+    }
+
+    void execute(std::vector<std::string> args, RTLIL::Design *design) YS_OVERRIDE
+    {
+        std::string run_from, run_to;
+        clear_flags();
+
+        size_t argidx;
+        for (argidx = 1; argidx < args.size(); argidx++)
+        {
+            if (args[argidx] == "-top" && argidx+1 < args.size()) {
+                top_opt = "-top " + args[++argidx];
+                continue;
+            }
+            if (args[argidx] == "-edif" && argidx+1 < args.size()) {
+                edif_file = args[++argidx];
+                continue;
+            }
+            if (args[argidx] == "-blif" && argidx+1 < args.size()) {
+                blif_file = args[++argidx];
+                continue;
+            }
+            if (args[argidx] == "-flatten") {
+                flatten = true;
+                continue;
+            }
+            break;
+        }
+        extra_args(args, argidx, design);
+
+        if (!design->full_selection())
+            log_cmd_error("This command only operates on fully selected designs!\n");
+
+        log_header(design, "Executing SYNTH_QUICKLOGIC pass.\n");
+        log_push();
+
+        run_script(design, run_from, run_to);
+
+        log_pop();
+    }
+
+    void script() YS_OVERRIDE
+    {
+        if (check_label("begin")) {
+            run("read_verilog -lib +/quicklogic/cells_sim.v");
+            run(stringf("hierarchy -check %s", (top_opt.empty()) ? "-auto-top" : top_opt.c_str()));
+        }
+
+        if (check_label("prepare")) {
+            run("proc");
+            if (flatten || help_mode)
+                run("flatten", "(with '-flatten')");
+        }
+
+        if (check_label("coarse")) {
+            run("tribuf -logic");
+            run("deminout");
+            run("opt");
+            run("opt_clean");
+            run("check");
+            run("peepopt");
+            run("opt_clean");
+            run("check");
+            run("techmap");
+            run("abc -lut 1:4");
```

Comment:
This looks *very* wrong. From the `abc` help:
```
    -lut <w1>:<w2>
        generate netlist using luts of (max) the specified width <w2>. All
        luts with width <= <w1> have constant cost. for luts larger than <w1>
        the area cost doubles with each additional input bit. the delay cost
        is still constant for all lut widths.
```

What this means is that a LUT2 costs twice as much area as a LUT1, a LUT3 four times as much area as a LUT1 and a LUT4 eight times as much area as a LUT1. This doesn't match the hardware at all: reading the EOS S3 datasheet, a Logic Cell can implement 2 LUT3s, a LUT4, or use the hard multiplexers to mux up to larger multiplexers. That means a LUT1, a LUT2 and LUT3 all have the same area cost.

You could consider either `abc -lut 3:4` to aim to reduce area by producing LUT3s that might be packed together into an LC; `abc -lut 4` to produce the fastest logic that fits in single LCs, or `abc -lut 4:6` to consider packing together multiple LCs to reduce logic depth. But `abc -lut 1:4` is surely wrong.

Reply:
You should also use `-dress` to attempt to preserve wire names in the netlist.

Reply:
I have switched to `abc -luts 1,2,2,4` also added MUX4 and MUX8 inference (this gives additional optimization for bigger desgins)

The problem I see is how to tell yosys that LUT1/MUX2 can be packed together with LUT2/3 pair or a single LUT4. @ZirconiumX any hints?

Reply:
I don't think that's presently possible. I think it would be better to let ABC just have the MUX2s, to see if it can perform logic optimisations on it, and then just pack as best as you can.

Reply:
@ZirconiumX can we land this PR and work on further optimizations later?